Project/Core/Logger.cs

```diff
@@ -1,113 +1,133 @@
-        {
-            using (System.IO.StreamWriter streamWriter = logFileInfo.AppendText())
-            {
-                streamWriter.Write(string.Format(Configuration.AppConfig.LOG_TEMPLATE_FORMAT, System.DateTime.Now, logMode, message));
-            }
-        }
-        /// <summary>
-        /// Writes a log to a file 
-        /// </summary>
-        /// <param name="logMode">
-        /// A log level
-        /// </param>
-        /// <param name="message">
-        /// Log message
-        /// </param>
-        [System.Obsolete("Core.Logger.Log has been deprecated. Please use other method Core.Logger.LogAsync", false)]
-        public void Log(LogMode logMode, string message)
-        {
-            try
-            {
-                writeLock.Wait();
-
-                CreateDirectoryIfNotExist();
-
-                DeleteLogFileIfBig();
-
-                AddMessageToLogFile(logMode, message);
-            }
-            finally
-            {
-                writeLock.Release();
-            }
-        }
-        /// <summary>
-        /// Writes a log to a file asynchronously
-        /// </summary>
-        /// <param name="logMode">
-        /// A log level
-        /// </param>
-        /// <param name="message">
-        /// Log message
-        /// </param>
-        public async void LogAsync(LogMode logMode, string message)
-        {
-            try
-            {
-                await writeLock.WaitAsync();
-
-                await System.Threading.Tasks.Task.Run(() => CreateDirectoryIfNotExist());
-
-                await System.Threading.Tasks.Task.Run(() => DeleteLogFileIfBig());
-
-                await System.Threading.Tasks.Task.Run(() => AddMessageToLogFile(logMode, message));
-            }
-            finally
-            {
-                writeLock.Release();
-            }
-        }
-    }
-}
+using System.Threading;
+
+namespace Core
+{
+    /// <summary>
+    /// Provides opportunities to log messages
+    /// <para/>
+    /// Implements a Singleton pattern
+    /// </summary>
+    public class Logger
+    {
+        // FIELDS
+        private readonly SemaphoreSlim writeLock; // because writing to file occupy process
+        private static Logger instance;
+        private LogMode offLogMode;
+        private readonly System.IO.FileInfo logFileInfo;
+
+        // CONSTRUCTORS
+        private Logger()
+        {
+            writeLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);
+            logFileInfo = new System.IO.FileInfo(Configuration.AppConfig.LOG_FILE);
+
+            // To turn off log mode(s) for all application pass it(them) to method "Off".
+            // For example, next line turns off Debug mode:
+            // Off(LogMode.Debug);
+            // For multiple off modes use "|" between them, next line turns off Debug and Info modes:
+            // Off(LogMode.Debug | LogMode.Info);
+            // To turn on modes that are off, only delete method "Off" with them.
```

Comment:
still bad implementaion. **On** method is missing. There is no oportnity to configurate Logger during runtime. Also why explanation to method is in private constructor? nobody exept developers can read it